Data/Hashable.hs

```diff
@@ -203,3 +209,25 @@ import Data.Hashable.Generic ()
 -- >                                 (1::Int) `hashWithSalt` n
 -- >     hashWithSalt s (Months n) = s `hashWithSalt`
 -- >                                 (2::Int) `hashWithSalt` n
+
+-- | A hashable value along with the result of the 'hash' function.
+data Hashed a = Hashed a {-# UNPACK #-} !Int
+                deriving Typeable
+
+-- | Wrap a hashable value, caching the 'hash' function result.
+hashed :: Hashable a => a -> Hashed a
+hashed a = Hashed a (hash a)
+
+-- | Unwrap hashed value.
+unhashed :: Hashed a -> a
+unhashed (Hashed a _) = a
+
+instance Eq a => Eq (Hashed a) where
+  Hashed a _ == Hashed b _ = a == b
+
+instance Ord a => Ord (Hashed a) where
+  Hashed a _ `compare` Hashed b _ = a `compare` b
+
+instance Hashable a => Hashable (Hashed a) where
+  hashWithSalt s = hashWithSalt s . unhashed
```

Comment:
We should define both `hash` and `hashWithSalt`. The latter could use `defaultHashWithSalt`.